### PR TITLE
PMM-11272 Add API for getting advisors

### PR DESCRIPTION
**Pull request by artemgavrilov**
PMM-11272

Build: SUBMODULES-3068

- [x] https://github.com/percona-platform/platform/pull/628
- [x] https://github.com/percona-platform/checked/pull/494

### CODE REVIEW

File: api/managementpb/checks.proto [hunk 1]
```diff
@@ -214,7 +235,18 @@ service SecurityChecks {
...
     option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
       summary: "List Security Checks"
-      description: "Returns a list of available Security Thread Tool checks."
+      description: "Returns a list of available checks."
+    };
+  }
+  // ListAdvisors returns a list of available Security Thread Tool checks.
+  rpc ListAdvisors(ListAdvisorsRequest) returns (ListAdvisorsResponse) {
+    option (google.api.http) = {
+      post: "/v1/management/Advisors/List"
```
Comment (artemgavrilov) [inline api/managementpb/checks.proto:244]: Still thinking about path

File: go.mod [hunk 2]
```diff
@@ -50,7 +50,7 @@ require (
 	github.com/minio/minio-go/v7 v7.0.48
 	github.com/mwitkow/go-proto-validators v0.3.2
 	github.com/percona-platform/dbaas-api v0.0.0-20221214134301-b6a331250826
-	github.com/percona-platform/saas v0.0.0-20221014123257-4fa7a15ce672
+	github.com/percona-platform/saas v0.0.0-20230214091511-a4450f505326
```
Comment (artemgavrilov) [inline go.mod:53]: TODO: Switch to main

File: managed/services/management/checks_test.go [hunk 3]
```diff
@@ -460,3 +461,48 @@ func TestUpdateSecurityChecks(t *testing.T) {
 		assert.Equal(t, &managementpb.ChangeSecurityChecksResponse{}, resp)
 	})
 }
+
+func TestCreateComment(t *testing.T) {
+	t.Parallel()
+
+	l := logrus.WithField("component", "tests")
+
+	testCases := []struct {
```
Comment (michal-kralik) [inline managed/services/management/checks_test.go:470]: What is the expected behavior if there are no technologies supported?

File: managed/services/utils_test.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+		"PostgreSQLSelect":         {Name: "PostgreSQLSelect", Version: 1, Type: check.PostgreSQLSelect},
+		"MongoDBGetParameter":      {Name: "MongoDBGetParameter", Version: 1, Type: check.MongoDBGetParameter},
+		"MongoDBBuildInfo":         {Name: "MongoDBBuildInfo", Version: 1, Type: check.MongoDBBuildInfo},
+		"MongoDBGetCmdLineOpts":    {Name: "MongoDBGetCmdLineOpts", Version: 1, Type: check.MongoDBGetCmdLineOpts},
+		"MongoDBReplSetGetStatus":  {Name: "MongoDBReplSetGetStatus", Version: 1, Type: check.MongoDBReplSetGetStatus},
+		"MongoDBGetDiagnosticData": {Name: "MongoDBGetDiagnosticData", Version: 1, Type: check.MongoDBGetDiagnosticData},
+		"unsupported type":         {Name: "unsupported type", Version: 1, Type: check.Type("RedisInfo")},
+		"missing type":             {Name: "missing type", Version: 1},
+		"MySQL family V2":          {Name: "MySQL family V2", Version: 2, Family: check.MySQL},
+		"PostrgeSQL family V2":     {Name: "PostrgeSQL family V2", Version: 2, Family: check.PostgreSQL},
```
Comment (michal-kralik) [inline managed/services/utils_test.go:43]:
```suggestion
		"PostgreSQL family V2":     {Name: "PostgreSQL family V2", Version: 2, Family: check.PostgreSQL},
```

**Review by michal-kralik — CHANGES_REQUESTED** [review]
looks good, only minor comments

File: go.mod [hunk 5]
```diff
@@ -2,6 +2,9 @@ module github.com/percona/pmm
 
 go 1.18
 
+// Update saas with
+// go get -v github.com/percona-platform/saas@latest
+
```
[the remark this answers is not in the document]
Comment (artemgavrilov) [reply go.mod:7]: Nope, we had this command here before. Probably it was dropped accidentally. This is often operation, it's useful to have it here.

File: api/managementpb/checks.proto [hunk 6]
```diff
@@ -214,7 +235,18 @@ service SecurityChecks {
...
+      body: "*"
+    };
+    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
+      summary: "List advisors"
+      description: "Returns a list of available advisors."
```
[the remark this answers is not in the document]
Comment (artemgavrilov) [reply api/managementpb/checks.proto:249]: That means available for the user. This endpoint returns different sets for different users depends on their plan.

File: managed/services/management/checks_test.go (hunk 3 above)
Comment (artemgavrilov) [reply managed/services/management/checks_test.go:470]: Every advisor has at least one check, so at least one technology is supported.

File: managed/services/checks/checks.go [hunk 7]
```diff
@@ -716,6 +743,7 @@ func (s *Service) executeChecksForTargetType(ctx context.Context, serviceType mo
 		}
 
 		for _, target := range targets {
+
```
[the remark this answers is not in the document]
Comment (artemgavrilov) [reply managed/services/checks/checks.go]: Done

File: api/managementpb/checks.proto [hunk 8]
```diff
@@ -214,7 +235,18 @@ service SecurityChecks {
     };
     option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
       summary: "List Security Checks"
-      description: "Returns a list of available Security Thread Tool checks."
+      description: "Returns a list of available checks."
+    };
+  }
+  // ListAdvisors returns a list of available Security Thread Tool checks.
```
[the remark this answers is not in the document]
Comment (artemgavrilov) [reply api/managementpb/checks.proto]: Yep, updated

File: managed/services/checks/checks.go [hunk 9]
```diff
@@ -1391,30 +1378,36 @@ func (s *Service) loadLocalChecks(file string) ([]check.Check, error) {
...
 	}
-	checks, err := check.Parse(bytes.NewReader(data), params)
+	checks, err := check.ParseChecks(bytes.NewReader(data), params)
 	if err != nil {
 		return nil, errors.Wrap(err, "failed to parse test checks file")
 	}
 
+	for _, c := range checks {
+		if c.Advisor != "dev" {
+			return nil, errors.Errorf("Local checks supposed to be linked to 'dev' advisor.")
```
[the remark this answers is not in the document]
Comment (artemgavrilov) [reply managed/services/checks/checks.go]: Done

File: managed/services/utils_test.go (hunk 4 above)
Comment (artemgavrilov) [reply managed/services/utils_test.go]: Fixed

File: go.mod (hunk 5 above)
[the remark this answers is not in the document]
Comment (michal-kralik) [reply go.mod:7]: I never used it myself. Maybe others didn't either ;)

File: api/managementpb/checks.proto (hunk 6 above)
[the remark this answers is not in the document]
Comment (michal-kralik) [reply api/managementpb/checks.proto:249]: I suggest we make this clearer in the description.

File: go.mod (hunk 5 above)
[the remark this answers is not in the document]
Comment (artemgavrilov) [reply go.mod:7]: I mean it's used when you develop something that affects portal and pmm at the same time. We have similar comments in every repo in portal. In this repo we have Portal API definitions and common entities. For example `advisors` defined there and during development you need to update this lib multiple times.

I thought that it should not bother anyone, but will help a little bit during development)

File: api/managementpb/checks.proto (hunk 6 above)
[the remark this answers is not in the document]
Comment (artemgavrilov) [reply api/managementpb/checks.proto]: Changed, check it out

File: api/managementpb/checks.proto [hunk 10]
```diff
@@ -206,15 +227,26 @@ service SecurityChecks {
       description: "Executes Security Thread Tool checks and returns when all checks are executed. All available checks will be started if check names aren't specified."
     };
   }
-  // ListSecurityChecks returns a list of available Security Thread Tool checks.
+  // ListSecurityChecks returns a list of advisor checks available for the user..
```
Comment (michal-kralik) [inline api/managementpb/checks.proto:230]:
```suggestion
  // ListSecurityChecks returns a list of advisor checks available to the user..
```

File: api/managementpb/checks.proto [hunk 11]
```diff
@@ -206,15 +227,26 @@ service SecurityChecks {
...
   rpc ListSecurityChecks(ListSecurityChecksRequest) returns (ListSecurityChecksResponse) {
     option (google.api.http) = {
       post: "/v1/management/SecurityChecks/List"
       body: "*"
     };
     option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
-      summary: "List Security Checks"
-      description: "Returns a list of available Security Thread Tool checks."
+      summary: "List advisor checks"
+      description: "Returns a list of checks available for the user."
```
Comment (michal-kralik) [inline api/managementpb/checks.proto:238]:
```suggestion
      description: "Returns a list of checks available to the user."
```

File: api/managementpb/checks.proto [hunk 12]
```diff
@@ -206,15 +227,26 @@ service SecurityChecks {
...
+  }
+  // ListAdvisors returns a list of advisors available for the user.
+  rpc ListAdvisors(ListAdvisorsRequest) returns (ListAdvisorsResponse) {
+    option (google.api.http) = {
+      post: "/v1/management/Advisors/List"
+      body: "*"
+    };
+    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
+      summary: "List advisors"
+      description: "Returns a list of advisors available for the user."
```
Comment (michal-kralik) [inline api/managementpb/checks.proto:249]:
```suggestion
      description: "Returns a list of advisors available to the user."
```

**Review by michal-kralik — APPROVED** [review]

File: managed/services/checks/checks_test.go [hunk 13]
```diff
@@ -716,7 +743,7 @@ func TestGetFailedChecks(t *testing.T) {
 
 		results, err := s.GetChecksResults(ctx, "test_svc")
 		assert.Nil(t, results)
-		assert.EqualError(t, err, services.ErrSTTDisabled.Error())
+		assert.EqualError(t, err, services.ErrAdvisorsDisabled.Error())
```
Comment (PavelKhripkov) [inline managed/services/checks/checks_test.go:746]:
```suggestion
		assert.ErrorIs(t, err, services.ErrAdvisorsDisabled)
```

File: managed/services/management/checks.go [hunk 14]
```diff
@@ -247,6 +248,75 @@ func (s *ChecksAPIService) ListSecurityChecks(_ context.Context, _ *managementpb
...
+	if len(mongoDBChecks) != 0 {
+		b = append(b, "MongoDB")
+	}
+
+	if len(b) == 3 {
+		return "All technologies supported"
+	}
+
+	return "Partial support (" + strings.Join(b, ", ") + ")"
+}
```
[the remark this answers is not in the document]
Comment (artemgavrilov) [reply managed/services/management/checks.go:318]: We need to show on the UI field that shows what technologies are supported by advisor. 
image

**Review by idoqo — APPROVED** [review]

File: managed/services/management/checks.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (PavelKhripkov) [reply managed/services/management/checks.go]: I started a discussion about the content of the field in slack. Other things were clarified in a call. So we can consider question as resolved.

**Review by PavelKhripkov — APPROVED** [review]

**Pull request merged**